In [1]:
import pandas as pd

In [75]:
df = pd.read_csv("Training_set1.csv")

#### Data Analysis

In [76]:
pd.set_option('display.max_columns',None)  # In order to see all the column names

In [77]:
df.head(10)

duration protocoltype      service flag  srcbytes  dstbytes  land  \
0         0          tcp  netbios_dgm  REJ         0         0     0   
1         0          tcp         smtp   SF      1239       400     0   
2         0          tcp         http   SF       222       945     0   
3         0          tcp         http   SF       235      1380     0   
4         0          tcp    uucp_path  REJ         0         0     0   
5         0          tcp         http   SF       348      2029     0   
6         0          tcp      private   S0         0         0     0   
7         0          tcp     csnet_ns   S0         0         0     0   
8         0          tcp    hostnames  REJ         0         0     0   
9         0          tcp         http   SF       242       380     0   

   wrongfragment  urgent  hot  numfailedlogins  loggedin  numcompromised  \
0              0       0    0                0         0               0   
1              0       0    0                0         1               0   
2              0       0    0                0         1               0   
3              0       0    0                0         1               0   
4              0       0    0                0         0               0   
5              0       0    0                0         1               0   
6              0       0    0                0         0               0   
7              0       0    0                0         0               0   
8              0       0    0                0         0               0   
9              0       0    0                0         1               0   

   rootshell  suattempted  numroot  numfilecreations  numshells  \
0          0            0        0                 0          0   
1          0            0        0                 0          0   
2          0            0        0                 0          0   
3          0            0        0                 0          0   
4          0            0        0                 0          0   
5          0            0        0                 0          0   
6          0            0        0                 0          0   
7          0            0        0                 0          0   
8          0            0        0                 0          0   
9          0            0        0                 0          0   

   numaccessfiles  numoutboundcmds  ishostlogin  isguestlogin  count  \
0               0                0            0             0    277   
1               0                0            0             0      1   
2               0                0            0             0     21   
3               0                0            0             0      4   
4               0                0            0             0    256   
5               0                0            0             0     27   
6               0                0            0             0    129   
7               0                0            0             0    129   
8               0                0            0             0    146   
9               0                0            0             0      5   

   srvcount  serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  \
0        15         0.0            0.0         1.0            1.0   
1         1         0.0            0.0         0.0            0.0   
2        25         0.0            0.0         0.0            0.0   
3         4         0.0            0.0         0.0            0.0   
4         3         0.0            0.0         1.0            1.0   
5        28         0.0            0.0         0.0            0.0   
6         7         1.0            1.0         0.0            0.0   
7        20         1.0            1.0         0.0            0.0   
8        20         0.0            0.0         1.0            1.0   
9         5         0.0            0.0         0.0            0.0   

   samesrvrate  diffsrvrate  srvdiffhostrate  dsthostcount  dsthostsrvcount  \
0  

In [5]:
df.shape

(86845, 43)

In [6]:
df.info() # Analysing the datatype of each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86845 entries, 0 to 86844
Data columns (total 43 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   duration                86845 non-null  int64  
 1   protocoltype            86845 non-null  object 
 2   service                 86845 non-null  object 
 3   flag                    86845 non-null  object 
 4   srcbytes                86845 non-null  int64  
 5   dstbytes                86845 non-null  int64  
 6   land                    86845 non-null  int64  
 7   wrongfragment           86845 non-null  int64  
 8   urgent                  86845 non-null  int64  
 9   hot                     86845 non-null  int64  
 10  numfailedlogins         86845 non-null  int64  
 11  loggedin                86845 non-null  int64  
 12  numcompromised          86845 non-null  int64  
 13  rootshell               86845 non-null  int64  
 14  suattempted             86845 non-null

In [7]:
df.isnull().sum()  #Finding the number of missing values in each column

duration                  0
protocoltype              0
service                   0
flag                      0
srcbytes                  0
dstbytes                  0
land                      0
wrongfragment             0
urgent                    0
hot                       0
numfailedlogins           0
loggedin                  0
numcompromised            0
rootshell                 0
suattempted               0
numroot                   0
numfilecreations          0
numshells                 0
numaccessfiles            0
numoutboundcmds           0
ishostlogin               0
isguestlogin              0
count                     0
srvcount                  0
serrorrate                0
srvserrorrate             0
rerrorrate                0
srvrerrorrate             0
samesrvrate               0
diffsrvrate               0
srvdiffhostrate           0
dsthostcount              0
dsthostsrvcount           0
dsthostsamesrvrate        0
dsthostdiffsrvrate        0
dsthostsamesrcportra

#### We have 3 categorical datatype columns, Hence dealing with categorical data as follows:
The Protocol Type column has just 3 categories hence, using One Hot Encoding 

In [8]:
df.protocoltype

0        tcp
1        tcp
2        tcp
3        tcp
4        tcp
        ... 
86840    tcp
86841    tcp
86842    tcp
86843    tcp
86844    tcp
Name: protocoltype, Length: 86845, dtype: object

In [9]:
df.protocoltype.value_counts()

protocoltype
tcp     75876
udp      9929
icmp     1040
Name: count, dtype: int64

In [10]:
df=pd.get_dummies(df,columns=['protocoltype'], prefix=['protocol'])  # Since their are only 3 unique values in this column, One hot encoding is used

In [11]:
df["protocol_icmp"]=df["protocol_icmp"].map({True:1,False:0}) # Mapping the boolean values to 1 and 0 respectively
df["protocol_tcp"]=df["protocol_tcp"].map({True:1,False:0})
df["protocol_udp"]=df["protocol_udp"].map({True:1,False:0})

In [12]:
df

duration      service flag  srcbytes  dstbytes  land  wrongfragment  \
0             0  netbios_dgm  REJ         0         0     0              0   
1             0         smtp   SF      1239       400     0              0   
2             0         http   SF       222       945     0              0   
3             0         http   SF       235      1380     0              0   
4             0    uucp_path  REJ         0         0     0              0   
...         ...          ...  ...       ...       ...   ...            ...   
86840         0         http   SF       216       674     0              0   
86841         0         http   SF       368      1725     0              0   
86842        28          ftp   SF      1483      4152     0              0   
86843         0         smtp   SF      1679       385     0              0   
86844     11343       telnet   SF      3396     16694     0              0   

       urgent  hot  numfailedlogins  loggedin  numcompromised  rootshell  \
0           0    0                0         0               0          0   
1           0    0                0         1               0          0   
2           0    0                0         1               0          0   
3           0    0                0         1               0          0   
4           0    0                0         0               0          0   
...       ...  ...              ...       ...             ...        ...   
86840       0    0                0         1               0          0   
86841       0    0                0         1               0          0   
86842       0   30                0         1               0          0   
86843       0    0                0         1               0          0   
86844       0    0                0         1               0          0   

       suattempted  numroot  numfilecreations  numshells  numaccessfiles  \
0                0        0                 0          0               0   
1                0        0                 0          0               0   
2                0        0                 0          0               0   
3                0        0                 0          0               0   
4                0        0                 0          0               0   
...            ...      ...               ...        ...             ...   
86840            0        0                 0          0               0   
86841            0        0                 0          0               0   
86842            0        0                 0          0               0   
86843            0        0                 0          0               0   
86844            0        0                 3          0               0   

       numoutboundcmds  ishostlogin  isguestlogin  count  srvcount  \
0                    0            0             0    277        15   
1                    0            0             0      1         1   
2                    0            0             0     21        25   
3                    0            0             0      4         4   
4                    0            0             0    256         3   
...                ...          ...           ...    ...       ...   
86840                0            0             0      2         2   
86841                0            0             0     19        19   
86842                0            0             1      1         1   
86843                0            0             0      1         1   
86844                0            0             0      1         1   

       serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  samesrvrate  \
0             0.0            0.0         1.0            1.0         0.05   
1             0.0            0.0         0.0            0.0         1.00   
2             0.0            0.0         0.0            0.0         1.00   
3             0.0            0.0         0.0            0.0         1.00   
4             0.0      

In [13]:
df.shape # no of columns has increased

(86845, 45)

In [14]:
pd.set_option('display.max_rows',70)

In [15]:
df.service.value_counts()

service
http           31505
private        12859
domain_u        7196
smtp            5806
ftp_data        4929
other           2131
telnet          1780
finger          1341
ftp             1134
auth             758
Z39_50           678
uucp             626
courier          583
uucp_path        556
bgp              554
whois            544
iso_tsap         534
time             516
nnsp             511
imap4            494
vmnet            492
urp_i            477
domain           445
supdup           428
csnet_ns         423
discard          423
ctf              422
http_443         415
eco_i            401
daytime          389
exec             381
gopher           376
efs              374
systat           361
name             353
link             350
hostnames        349
login            346
klogin           345
mtp              335
echo             334
ldap             315
netbios_dgm      310
sunrpc           303
netbios_ssn      289
netstat          282
netbios_ns       268
ssh  

In [16]:
df.service.nunique()

65

 Here there are 65 unique values hence using Mean Ordinal Encoding

In [17]:
Mean_of_services= df.groupby(['service'])['attack'].mean().to_dict()
# Calculating the mean of the Target variable according to the Service and converting it to a Dictionary to fit into the map function

In [18]:
Mean_of_services

{'IRC': 0.0,
 'X11': 0.0,
 'Z39_50': 1.0,
 'auth': 0.7440633245382586,
 'bgp': 1.0,
 'courier': 1.0,
 'csnet_ns': 1.0,
 'ctf': 1.0,
 'daytime': 1.0,
 'discard': 1.0,
 'domain': 0.9258426966292135,
 'domain_u': 0.0,
 'echo': 1.0,
 'eco_i': 0.0,
 'ecr_i': 0.0,
 'efs': 1.0,
 'exec': 1.0,
 'finger': 0.6726323639075317,
 'ftp': 0.35008818342151676,
 'ftp_data': 0.19314262527896126,
 'gopher': 1.0,
 'hostnames': 1.0,
 'http': 0.03351848912870973,
 'http_443': 1.0,
 'imap4': 0.9939271255060729,
 'iso_tsap': 1.0,
 'klogin': 1.0,
 'kshell': 1.0,
 'ldap': 1.0,
 'link': 1.0,
 'login': 1.0,
 'mtp': 1.0,
 'name': 1.0,
 'netbios_dgm': 1.0,
 'netbios_ns': 1.0,
 'netbios_ssn': 1.0,
 'netstat': 1.0,
 'nnsp': 1.0,
 'nntp': 1.0,
 'ntp_u': 0.0,
 'other': 0.02111684655091506,
 'pop_2': 1.0,
 'pop_3': 0.2564102564102564,
 'printer': 1.0,
 'private': 0.938253363403064,
 'red_i': 0.0,
 'remote_job': 1.0,
 'rje': 1.0,
 'shell': 0.9387755102040817,
 'smtp': 0.03341370995521874,
 'sql_net': 1.0,
 'ssh': 0.987234

In [19]:
df['service'] =df['service'].map(Mean_of_services)

In [20]:
df['service']

0        1.000000
1        0.033414
2        0.033518
3        0.033518
4        1.000000
           ...   
86840    0.033518
86841    0.033518
86842    0.350088
86843    0.033414
86844    0.581461
Name: service, Length: 86845, dtype: float64

In [21]:
df

duration   service flag  srcbytes  dstbytes  land  wrongfragment  \
0             0  1.000000  REJ         0         0     0              0   
1             0  0.033414   SF      1239       400     0              0   
2             0  0.033518   SF       222       945     0              0   
3             0  0.033518   SF       235      1380     0              0   
4             0  1.000000  REJ         0         0     0              0   
...         ...       ...  ...       ...       ...   ...            ...   
86840         0  0.033518   SF       216       674     0              0   
86841         0  0.033518   SF       368      1725     0              0   
86842        28  0.350088   SF      1483      4152     0              0   
86843         0  0.033414   SF      1679       385     0              0   
86844     11343  0.581461   SF      3396     16694     0              0   

       urgent  hot  numfailedlogins  loggedin  numcompromised  rootshell  \
0           0    0                0         0               0          0   
1           0    0                0         1               0          0   
2           0    0                0         1               0          0   
3           0    0                0         1               0          0   
4           0    0                0         0               0          0   
...       ...  ...              ...       ...             ...        ...   
86840       0    0                0         1               0          0   
86841       0    0                0         1               0          0   
86842       0   30                0         1               0          0   
86843       0    0                0         1               0          0   
86844       0    0                0         1               0          0   

       suattempted  numroot  numfilecreations  numshells  numaccessfiles  \
0                0        0                 0          0               0   
1                0        0                 0          0               0   
2                0        0                 0          0               0   
3                0        0                 0          0               0   
4                0        0                 0          0               0   
...            ...      ...               ...        ...             ...   
86840            0        0                 0          0               0   
86841            0        0                 0          0               0   
86842            0        0                 0          0               0   
86843            0        0                 0          0               0   
86844            0        0                 3          0               0   

       numoutboundcmds  ishostlogin  isguestlogin  count  srvcount  \
0                    0            0             0    277        15   
1                    0            0             0      1         1   
2                    0            0             0     21        25   
3                    0            0             0      4         4   
4                    0            0             0    256         3   
...                ...          ...           ...    ...       ...   
86840                0            0             0      2         2   
86841                0            0             0     19        19   
86842                0            0             1      1         1   
86843                0            0             0      1         1   
86844                0            0             0      1         1   

       serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  samesrvrate  \
0             0.0            0.0         1.0            1.0         0.05   
1             0.0            0.0         0.0            0.0         1.00   
2             0.0            0.0         0.0            0.0         1.00   
3             0.0            0.0         0.0            0.0         1.00   
4             0.0            0.0         1.0            1.0

In [22]:
df.shape

(86845, 45)

In [23]:
df.flag.value_counts()

flag
SF      50672
S0      27773
REJ      6721
RSTO     1127
S1        294
RSTR      106
S2        103
S3         38
OTH        10
SH          1
Name: count, dtype: int64

Using Label Encoding since their are just 10 unique values

In [24]:
flag_convert={"SF":1,"S0":2,"REJ":3,"RSTO":4,'S1':5, "RSTR":6,'S2':7,'S3':8,"OTH":9,'SH':10}

In [25]:
df['flag']=df.flag.map(flag_convert)

In [26]:
df['flag']

0        3
1        1
2        1
3        1
4        3
        ..
86840    1
86841    1
86842    1
86843    1
86844    1
Name: flag, Length: 86845, dtype: int64

In [27]:
df

duration   service  flag  srcbytes  dstbytes  land  wrongfragment  \
0             0  1.000000     3         0         0     0              0   
1             0  0.033414     1      1239       400     0              0   
2             0  0.033518     1       222       945     0              0   
3             0  0.033518     1       235      1380     0              0   
4             0  1.000000     3         0         0     0              0   
...         ...       ...   ...       ...       ...   ...            ...   
86840         0  0.033518     1       216       674     0              0   
86841         0  0.033518     1       368      1725     0              0   
86842        28  0.350088     1      1483      4152     0              0   
86843         0  0.033414     1      1679       385     0              0   
86844     11343  0.581461     1      3396     16694     0              0   

       urgent  hot  numfailedlogins  loggedin  numcompromised  rootshell  \
0           0    0                0         0               0          0   
1           0    0                0         1               0          0   
2           0    0                0         1               0          0   
3           0    0                0         1               0          0   
4           0    0                0         0               0          0   
...       ...  ...              ...       ...             ...        ...   
86840       0    0                0         1               0          0   
86841       0    0                0         1               0          0   
86842       0   30                0         1               0          0   
86843       0    0                0         1               0          0   
86844       0    0                0         1               0          0   

       suattempted  numroot  numfilecreations  numshells  numaccessfiles  \
0                0        0                 0          0               0   
1                0        0                 0          0               0   
2                0        0                 0          0               0   
3                0        0                 0          0               0   
4                0        0                 0          0               0   
...            ...      ...               ...        ...             ...   
86840            0        0                 0          0               0   
86841            0        0                 0          0               0   
86842            0        0                 0          0               0   
86843            0        0                 0          0               0   
86844            0        0                 3          0               0   

       numoutboundcmds  ishostlogin  isguestlogin  count  srvcount  \
0                    0            0             0    277        15   
1                    0            0             0      1         1   
2                    0            0             0     21        25   
3                    0            0             0      4         4   
4                    0            0             0    256         3   
...                ...          ...           ...    ...       ...   
86840                0            0             0      2         2   
86841                0            0             0     19        19   
86842                0            0             1      1         1   
86843                0            0             0      1         1   
86844                0            0             0      1         1   

       serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  samesrvrate  \
0             0.0            0.0         1.0            1.0         0.05   
1             0.0            0.0         0.0            0.0         1.00   
2             0.0            0.0         0.0            0.0         1.00   
3             0.0            0.0         0.0            0.0         1.00   
4             0.0            0.0         1.0   

In [28]:
df.dtypes

duration                    int64
service                   float64
flag                        int64
srcbytes                    int64
dstbytes                    int64
land                        int64
wrongfragment               int64
urgent                      int64
hot                         int64
numfailedlogins             int64
loggedin                    int64
numcompromised              int64
rootshell                   int64
suattempted                 int64
numroot                     int64
numfilecreations            int64
numshells                   int64
numaccessfiles              int64
numoutboundcmds             int64
ishostlogin                 int64
isguestlogin                int64
count                       int64
srvcount                    int64
serrorrate                float64
srvserrorrate             float64
rerrorrate                float64
srvrerrorrate             float64
samesrvrate               float64
diffsrvrate               float64
srvdiffhostrat

#### All Categorical Values are now Converted to Int and Float Values
#### Scaling the Float and Int Values
Using MinMax Scaler

In [29]:
import sklearn 
import numpy as np

In [30]:
target = df.loc[:,['attack']]

In [31]:
target

attack
0           1
1           0
2           0
3           0
4           1
...       ...
86840       0
86841       0
86842       0
86843       0
86844       0

[86845 rows x 1 columns]

In [32]:
df.columns

Index(['duration', 'service', 'flag', 'srcbytes', 'dstbytes', 'land',
       'wrongfragment', 'urgent', 'hot', 'numfailedlogins', 'loggedin',
       'numcompromised', 'rootshell', 'suattempted', 'numroot',
       'numfilecreations', 'numshells', 'numaccessfiles', 'numoutboundcmds',
       'ishostlogin', 'isguestlogin', 'count', 'srvcount', 'serrorrate',
       'srvserrorrate', 'rerrorrate', 'srvrerrorrate', 'samesrvrate',
       'diffsrvrate', 'srvdiffhostrate', 'dsthostcount', 'dsthostsrvcount',
       'dsthostsamesrvrate', 'dsthostdiffsrvrate', 'dsthostsamesrcportrate',
       'dsthostsrvdiffhostrate', 'dsthostserrorrate', 'dsthostsrvserrorrate',
       'dsthostrerrorrate', 'dsthostsrvrerrorrate', 'lastflag', 'attack',
       'protocol_icmp', 'protocol_tcp', 'protocol_udp'],
      dtype='object')

In [34]:
training_columns=df.loc[:,['duration', 'service', 'flag', 'srcbytes', 'dstbytes', 'land',
       'wrongfragment', 'urgent', 'hot', 'numfailedlogins', 'loggedin',
       'numcompromised', 'rootshell', 'suattempted', 'numroot',
       'numfilecreations', 'numshells', 'numaccessfiles', 'numoutboundcmds',
       'ishostlogin', 'isguestlogin', 'count', 'srvcount', 'serrorrate',
       'srvserrorrate', 'rerrorrate', 'srvrerrorrate', 'samesrvrate',
       'diffsrvrate', 'srvdiffhostrate', 'dsthostcount', 'dsthostsrvcount',
       'dsthostsamesrvrate', 'dsthostdiffsrvrate', 'dsthostsamesrcportrate',
       'dsthostsrvdiffhostrate', 'dsthostserrorrate', 'dsthostsrvserrorrate',
       'dsthostrerrorrate', 'dsthostsrvrerrorrate', 'lastflag',
       'protocol_icmp', 'protocol_tcp', 'protocol_udp']]

In [35]:
training_columns

duration   service  flag  srcbytes  dstbytes  land  wrongfragment  \
0             0  1.000000     3         0         0     0              0   
1             0  0.033414     1      1239       400     0              0   
2             0  0.033518     1       222       945     0              0   
3             0  0.033518     1       235      1380     0              0   
4             0  1.000000     3         0         0     0              0   
...         ...       ...   ...       ...       ...   ...            ...   
86840         0  0.033518     1       216       674     0              0   
86841         0  0.033518     1       368      1725     0              0   
86842        28  0.350088     1      1483      4152     0              0   
86843         0  0.033414     1      1679       385     0              0   
86844     11343  0.581461     1      3396     16694     0              0   

       urgent  hot  numfailedlogins  loggedin  numcompromised  rootshell  \
0           0    0                0         0               0          0   
1           0    0                0         1               0          0   
2           0    0                0         1               0          0   
3           0    0                0         1               0          0   
4           0    0                0         0               0          0   
...       ...  ...              ...       ...             ...        ...   
86840       0    0                0         1               0          0   
86841       0    0                0         1               0          0   
86842       0   30                0         1               0          0   
86843       0    0                0         1               0          0   
86844       0    0                0         1               0          0   

       suattempted  numroot  numfilecreations  numshells  numaccessfiles  \
0                0        0                 0          0               0   
1                0        0                 0          0               0   
2                0        0                 0          0               0   
3                0        0                 0          0               0   
4                0        0                 0          0               0   
...            ...      ...               ...        ...             ...   
86840            0        0                 0          0               0   
86841            0        0                 0          0               0   
86842            0        0                 0          0               0   
86843            0        0                 0          0               0   
86844            0        0                 3          0               0   

       numoutboundcmds  ishostlogin  isguestlogin  count  srvcount  \
0                    0            0             0    277        15   
1                    0            0             0      1         1   
2                    0            0             0     21        25   
3                    0            0             0      4         4   
4                    0            0             0    256         3   
...                ...          ...           ...    ...       ...   
86840                0            0             0      2         2   
86841                0            0             0     19        19   
86842                0            0             1      1         1   
86843                0            0             0      1         1   
86844                0            0             0      1         1   

       serrorrate  srvserrorrate  rerrorrate  srvrerrorrate  samesrvrate  \
0             0.0            0.0         1.0            1.0         0.05   
1             0.0            0.0         0.0            0.0         1.00   
2             0.0            0.0         0.0            0.0         1.00   
3             0.0            0.0         0.0            0.0         1.00   
4             0.0            0.0         1.0   

In [36]:
from sklearn.model_selection import train_test_split

In [37]:
training_column_train, val_training_column, target_train,val_target=train_test_split(training_columns,target,test_size = 0.3,random_state = 2)
target_train = np.ravel(target_train)
val_target= np.ravel(val_target)

In [38]:
from sklearn.preprocessing import MinMaxScaler

In [39]:
scaler=MinMaxScaler(feature_range = (0,1))

In [40]:
scaled_train_columns = scaler.fit_transform(training_column_train)

In [41]:
scaled_train_columns

array([[0.        , 0.03351849, 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.93825336, 0.11111111, ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.03351849, 0.        , ..., 0.        , 1.        ,
        0.        ],
       ...,
       [0.        , 0.93825336, 0.11111111, ..., 0.        , 1.        ,
        0.        ],
       [0.        , 1.        , 0.11111111, ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.03351849, 0.        , ..., 0.        , 1.        ,
        0.        ]])

In [42]:
from sklearn.linear_model import LogisticRegression

In [43]:
logmodel= LogisticRegression()

In [44]:
logmodel.fit(scaled_train_columns,target_train)

LogisticRegression()

In [45]:
val_training_scaled = scaler.transform(val_training_column)

In [46]:
logmodel.predict(val_training_scaled)

array([0, 0, 1, ..., 0, 1, 0])

In [47]:
from sklearn.metrics import mean_squared_error

In [48]:
mean_squared_error(logmodel.predict(val_training_scaled), val_target)

0.00038381822368926076

In [49]:
logmodel.score(scaled_train_columns,target_train)

0.999572305110954

In [50]:
logmodel.score(val_training_scaled,val_target)

0.9996161817763107

In [51]:
scaled_testdata = pd.read_csv("Scaled_Test_data.csv")

In [52]:
pedicted_value_of_attacks=logmodel.predict(scaled_testdata)

/opt/homebrew/Cellar/jupyterlab/4.1.6_1/libexec/lib/python3.12/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


In [53]:
pedicted_value_of_attacks

array([1, 0, 1, ..., 1, 0, 1])

In [54]:
pedicted_value_of_attacks_df = pd.DataFrame(pedicted_value_of_attacks,columns = ['attack'])

In [62]:
pedicted_value_of_attacks_df  # Final Solution to the Problem Statement

attack
0           1
1           0
2           1
3           1
4           1
...       ...
21707       0
21708       0
21709       1
21710       0
21711       1

[21712 rows x 1 columns]

In [56]:
pedicted_value_of_attacks_df.to_csv("Submission_AttackPrediction123.csv",index=False)

In [57]:
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay

In [64]:
confusion_matrix(y_true= val_target, y_pred=logmodel.predict(val_training_scaled))

array([[16213,     7],
       [    3,  9831]])

#### Here, the TN (True Negatives)= 16213, FP(False Positives)=7 , FN(False Negatives)=3, TP(True Positives)=9831

In [67]:
confusion_matrix(y_true= val_target, y_pred=logmodel.predict(val_training_scaled)).ravel()

array([16213,     7,     3,  9831])

In [72]:
TN,FP,FN,TP=confusion_matrix(y_true= val_target, y_pred=logmodel.predict(val_training_scaled)).ravel()
print(TN,FP,FN,TP)

16213 7 3 9831


In [73]:
from sklearn.metrics import f1_score

In [ ]:
f1_score(val_target,logmodel.predict(val_training_scaled))